app/src/main/java/org/openobservatory/ooniprobe/model/database/Url.java

```diff
@@ -56,6 +75,52 @@ public static Url checkExistingUrl(String input, String categoryCode, String cou
 		return url;
 	}
 
+
+	/**
+	 * This function saves or updates input urls ({@code List<Url>}) into the database and returns the
+	 * list of strings ({@code List<String>}) used internally to perform this operation, where each string
+	 * actually useful to continue processing the input URLs.
+	 *
+	 * @param urls
+	 *        List of urls to save.
+	 * @return list of url strings.
+	 */
+	public static List<String> saveOrUpdate(List<Url> urls) {
+
+		Map<String, Url> resultUrlsMap = Maps.uniqueIndex(
+				urls,
+				input -> input.url
+		);
+
+		List<Url> existingUrls = Url.getExistingUrls(resultUrlsMap.keySet());
+
+		List<Url> existingUrlsToUpdate = Lists.newArrayList(Iterables.filter(existingUrls, input -> {
+			Url incomingChanges = resultUrlsMap.get(input.url);
+			if (incomingChanges == null) {
+				return false;
+			}
+			return ((!input.category_code.equals(incomingChanges.category_code)
+					&& !incomingChanges.category_code.equals("MISC"))
+					|| (!input.country_code.equals(incomingChanges.country_code)
+					&& !incomingChanges.country_code.equals("XX")));
+		}));
+
+		if (!existingUrlsToUpdate.isEmpty()) {
+			Url.saveAll(Lists.newArrayList(existingUrlsToUpdate));
+		}
+
+		if (existingUrls.size() != resultUrlsMap.size()) {
+
+			List<String> existingUrlStrings = Lists.newArrayList(Iterables.transform(existingUrls, input -> input.url));
```

Comment:
It's going to be a bit faster when you do the lookup below if you store the existingUrls inside of a `Map`.

It's probably going to be just as fast and perhaps a bit more legible if instead of filtering through the maps you use a `HashMap` and do a set difference between the two.

I am thinking of something like:
```
Set<String> existingUrlSet = new HashSet<String>(Lists.newArrayList(Iterables.transform(existingUrls, input -> input.url)))
Set<String> newUrlSet = new HashSet<string>(resultUrlsMap.values())
newUrlSet.removeAll(existingUrlSet)
```